src/server/pfs/server/master.go

```diff
@@ -87,78 +93,134 @@ func (d *driver) master(ctx context.Context) {
+							return errors.Wrap(err, "watch repo event")
 						}
-						cancel()
-						delete(repos, key)
+					case <-ctx.Done():
+						return nil
 					}
-					return nil
 				}
-				if _, ok := repos[key]; ok {
+			})
+			// Get existing entries.
+			existingRepos := make([]*postgres.Event, 0)
+			if err := dbutil.WithTx(ctx, d.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+				iter, err := pfsdb.ListRepo(ctx, tx, nil)
+				if err != nil {
+					return errors.Wrap(err, "create list repo iterator")
+				}
+				return errors.Wrap(stream.ForEach[pfsdb.RepoPair](ctx, iter, func(repoPair pfsdb.RepoPair) error {
+					event := &postgres.Event{
+						Id:   uint64(repoPair.ID),
+						Type: postgres.EventInsert,
+					}
+					existingRepos = append(existingRepos, event)
 					return nil
+				}), "for each repo")
+			}, dbutil.WithReadOnly()); err != nil {
+				return errors.Wrap(err, "list repos")
+			}
+			for _, event := range existingRepos {
+				if err := d.manageRepos(ctx, ring, repos, event); err != nil {
+					return errors.Wrap(err, "manage repos for existing entries")
 				}
-				ctx, cancel := pctx.WithCancel(ctx)
-				repos[key] = cancel
-				go func() {
-					backoff.RetryUntilCancel(ctx, func() (retErr error) { //nolint:errcheck
-						ctx, cancel := pctx.WithCancel(ctx)
-						defer cancel()
-						var err error
-						ctx, err = ring.Lock(ctx, lockPrefix)
-						if err != nil {
-							return errors.Wrap(err, "error locking repo lock")
-						}
-						defer errors.Invoke1(&retErr, ring.Unlock, lockPrefix, "unlocking repo lock")
-						var eg errgroup.Group
-						eg.Go(func() error {
-							return backoff.RetryUntilCancel(ctx, func() error {
-								return d.manageBranches(ctx, key)
-							}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-								log.Error(ctx, "error managing branches", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-								return nil
-							})
-						})
-						eg.Go(func() error {
-							return backoff.RetryUntilCancel(ctx, func() error {
-								return d.finishRepoCommits(ctx, key)
-							}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-								log.Error(ctx, "error finishing repo commits", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-								return nil
-							})
-						})
-						return errors.EnsureStack(eg.Wait())
-					}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-						log.Error(ctx, "error managing repo", zap.String("repo", key), zap.Error(err), zap.Duration("retryAfter", d))
-						return nil
-					})
-				}()
-				return nil
-			})
-			return errors.EnsureStack(err)
+			}
+			return errors.Wrap(eg.Wait(), "waiting for manageRepos goroutines to finish")
 		},
 	)
 }
 
+func (d *driver) manageRepos(ctx context.Context, ring *consistenthashing.Ring, repos map[pfsdb.RepoID]context.CancelFunc, ev *postgres.Event) error {
```

Comment:
This can manage a single repo, and avoid needing the map.  Will be canceled through context by the main loop, when a delete occurs.